man/pfl.1
.\" Automatically generated by Pandoc 3.10.1
.\"
.TH "PFL" "1" "" "pfl user manual" "Version 3.6.0"
.SH NAME
pfl \- Upload installed packages info since last run to PFL to make it
searchable.
.SH SYNOPSIS
\f[B]pfl\f[R] [\-p|\(enpretend] [\-a \f[I]ATOM\f[R]|\(enatom
\f[I]ATOM\f[R]] [\-r \f[I]REPO\f[R]| \(enrepo \f[I]REPO\f[R]]
.PP
\f[B]pfl\f[R] [\-h] [\-v]
.SH DESCRIPTION
This is the PFL upload command.
The purpose of this command is to collect the file names (not the
content) of all installed packages from the Gentoo and GURU repo and
upload them to portagefilelist.de.
After some time your uploaded data will be imported into a searchable
database.
Thus you will provide a way for other people to find a package which
contains a specific file/binary/command.
Please visit https://www.portagefilelist.de for further information.
.SH USAGE
\f[B]pfl\f[R] [\-p|\-\-pretend] [\-a \f[I]ATOM\f[R]|\-\-atom
\f[I]ATOM\f[R]] [\-r \f[I]REPO\f[R]| \-\-repo \f[I]REPO\f[R]] [\-h]
[\-v]
.SH OPTIONS
.TP
\f[CR]\-a ATOM, \-\-atom ATOM\f[R]
Update only for given atom.
.TP
\f[CR]\-p, \-\-pretend\f[R]
Collect data only and do not upload or change the last run value.
.TP
\f[CR]\-r REPO, \-\-repo REPO\f[R]
Update only for given repository: gentoo|guru
.TP
\f[CR]\-h, \-\-help\f[R]
Show this help message and exit.
.TP
\f[CR]\-v, \-\-version\f[R]
Show version number and exit.
.SH EXIT STATUS
Nothing special yet.
A non\-zero exit code is treated as an abnormal exit, and at times, the
error code indicates what the problem was.
A zero error code means a successful exit.
.SH EXAMPLES
Usual execution as a user: \f[CR]pfl\f[R]
.PP
Do not upload but create the upload files to be viewed:
\f[CR]pfl \-p\f[R]
.PP
Update only for given repository.
Values are gentoo or guru: \f[CR]pfl \-r guru\f[R]
.PP
Update only for a given package atom:
\f[CR]pfl \-atom =media\-fonts/fira\-code\-6.2\f[R].
Only the specific package atom syntax is currently supported.
Like =media\-fonts/fira\-code\-6.2 and NOT media\-fonts/fira\-code
.SH FILES
.TP
\(ti/\f[B].pfl.info\f[R]
Created after first run and stores current pfl version and last
execution time.
Execution time is used to gather only packages since last run.
Remove file to reset last run time and gather all packages.
.TP
/var/lib/pfl/\f[B]pfl.info\f[R]
File used if run in system/cron context.
Created after first run and stores current pfl version and last
execution time.
Execution time is used to gather only packages since last run.
Remove file to reset last run time and gather all packages.
.SH BUG REPORTS
Report to https://github.com/portagefilelist/client/issue
.SH COPYRIGHT
2023\-2026 Banana mail\(atbananas\-playground.net
.PP
?\-2023 D. Buschke \- Original creator
.PP
License: GNU General Public License v2 (GPLv2)
.SH SEE ALSO
e\-file(1)
.SH VERSION
3.6.0


CHANGELOG
pfl-3.6.0 2026-09-05

+ Added new option to e-file. '-p', '--package', Search for packname instead.
+ Added python 3.13 and 3.14 notice to project.toml.
+ Fixed #65. Verbosity is too high

pfl-3.5.3 - 2024-05-18

+ pfl.service: use (After|Wants)=network-online.target. Thx to https://github.com/Flowdalic

pfl-3.5.2 - 2024-03-30

+ Fixed with PR #53: The lack of whitespace with listed versions. Thx to https://github.com/parona-source
+ Fixed and improved portage versions output #54
+ Fixed --atom option. It truncated the -rX from version numbers. Eg. 6.6.0-r3 became 6.6.0 and thus broke the lookup.
+ Added [--plain] option to e-file. Output without any fancy stuff, just plain text.
+ Changed output about installed or not. Installed version is returned every time now, but displays versions
  only if installed.
+ Cleared text formatting in readme and man files
+ Now handles the usr-merge topic https://wiki.gentoo.org/wiki/Merge-usr

pfl-3.5.1 - 2024-03-09

+ Fixed 51: Damn copy&pasta error. Sorry about that. Don't trust the .0 versions :-)

pfl-3.5 - 2024-03-07

+ Fixed #47: Do not process everything after a version change
+ Fixed https://github.com/portagefilelist/client/issues/42 with https://github.com/portagefilelist/client/pull/49
  from https://github.com/AndrewAmmerlaan
+ Better modules. It is one package (pfl) and two modules (e_file, pfl) with a run() method now.
+ man files are .md files now and will be converted with pandoc #44
+ Do not process everything after a version change
+ Improved terminal output format.

pfl-3.4 2024-01-06

+ Fixed man file for pfl
+ Added GURU as a additional repository #41
+ Added --repo [-r|--repo] cli option to pfl to collect and update only from specified repository.
  Only Gentoo and GURU are available.
+ Changed the upload file creation process to avoid big files.
  It creates new a file for each category, compress and adds them to a single tar file which is then uploaded.
  Backwards compatible to older versions.
+ Updated man files

pfl-3.3.1 2023-12-18

+ Cleared license confusion. Will be GPLv2

pfl-3.3 2023-12-18

+ Added pretend [-p|--pretend] cli option to pfl. Data will be collected to be viewed, but not uploaded. #33
+ Added the ability to pfl to provide a package atom [-a|--atom] to upload only the provided package. #13
+ Added man file for e-file
+ Updated description texts
+ Added CHANGELOG and VERSION
+ Move to setuptools and PEP517 (thx to https://github.com/parona-source)
+ Use EROOT for determining root (thx to https://github.com/parona-source)
+ Use fromtimestamp instead of utcfromtimestamp (thx to https://github.com/parona-source)
+ Updated license to GPLv3


pfl-3.2.1 2023-07-18

+ Under new management
+ Created pfl.1 by @JensKorte in #19
* Update README.md by @JensKorte in #17
* Update setup.py by @billie80 in #23
* Update e-file @bananas-repos


pfl-3.2 2021-10-06

+ python version of e-file (#7, #8)
+ Not using HOME (#10)
+ empty temp files not removed (#12)
+ run pfl script as user/group portage


pfl-3.1 2020-04-12

+ HTTPS support (#3)
+ IPv6 support (#2)
+ Documentation (#5)
+ systemd Timer (#4)
* wrong xml encoding in file names (#1)


 pfl-3.0.1-r1 2020-04-08

+ Init version pfl-3.0.1-r1


pfl/pfl.py
# Copyright 2024 Banana and contributors of https://github.com/portagefilelist
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# standard library
import sys
import os
import pwd
from tempfile import mkstemp, mkdtemp
from time import time
import configparser
from xml.sax.saxutils import escape
import tarfile

# external library
# portage api: sys-apps/portage
import portage
# http: dev-python/requests
import requests

VERSION = '3.6.0'
UPLOADURL='https://www.portagefilelist.de/data.php'
ALLOWED_REPOS = ['gentoo', 'guru']

# if it is run as cron and portage use. Otherwise use current user HOME
if pwd.getpwuid(os.getuid())[0] == 'portage':
    INFOFILE = '/var/lib/pfl/pfl.info'
else:
    HOME = os.path.expanduser("~")
    INFOFILE = '%s/.pfl.info' % HOME

# the main method to run this.
# options are
# options = {
#      'onlyRepo': '',
#      'onlyPackage': '',
#      'pretend': args.pretend,
#      'stdout': False
#  }
# Use options['stdout'] = True if you want to run this as a script which prints the output as it happens.
# With False the output is collected and returned, so no immediate display what is going on.
def run(options):
    start = PFL(options)
    return start.run()

# https://dev.gentoo.org/~zmedico/portage/doc/api/index.html
# https://wiki.gentoo.org/wiki/Project:Portage
class PortageMangle(object):
    _settings = None
    _vardbapi = None
    _options = None
    _out = ''

    def __init__(self, options):
        self._options = options
        eroot = portage.settings['EROOT']
        if eroot in portage.db:
            self._settings = portage.db[eroot]['vartree'].settings
            self._vardbapi = portage.db[eroot]['vartree'].dbapi
        else:
            raise Exception(f'Tree "{eroot}" not present.')

    def log(self, output):
        if self._options['stdout']:
            print(output)
        else:
            self._out += output + '\n'

    def get_wellknown_cpvs(self, since):
        if self._options['onlyPackage']:
            if self._vardbapi.cpv_exists(self._options['onlyPackage']):
                # category, package, version of specific package
                c, p, v, r = portage.versions.catpkgsplit(self._options['onlyPackage'])
                if r != 'r0':
                    v = '%s-%s' % (v, r)
                cpvs = [c + '/' + p + '-' + v]
            else:
                self.log('No such atom installed.')
                return None
        else:
            # category, package, version of all installed packages
            cpvs = self._vardbapi.cpv_all()

        # search for pkgs from known repositories
        wellknown = {}
        wellknown_count = 0
        for cpv in cpvs:
            # category, package, version, revision
            c, p, v, r = portage.versions.catpkgsplit(cpv)
            if r != 'r0':
                v = '%s-%s' % (v, r)

            # repository
            repo, = self._vardbapi.aux_get(cpv, ['repository'])
            if len(repo) == 0:
                repo, = self._vardbapi.aux_get(cpv, ['REPOSITORY'])

            # timestamp of merge
            mergedstamp = self._vardbapi.aux_get(cpv, ['_mtime_'])[0]
            # Some installs started to create float timestamp
            # I don't know why it was not float already all the time
            # Something in the creation of _mtime_ must have been changed
            mergedstamp = int(round(mergedstamp))

            if repo in ALLOWED_REPOS and mergedstamp >= since:
                if self._options['onlyRepo'] and repo != self._options['onlyRepo']:
                    continue

                wellknown.setdefault(repo, {}).setdefault(c, {}).setdefault(p, []).append(v)
                wellknown_count = wellknown_count + 1

        return [wellknown_count, wellknown]

    def get_contents(self, c, p, v):
        dbl = portage.dblink(c, '%s-%s' % (p, v), self._settings['ROOT'], self._settings)
        return dbl.getcontents()

    def _write2file(self, file, txt, indent):
        os.write(file[0], bytes(indent + txt, 'UTF-8'))

    # create a xml file per category and return a list of created xml files
    def collect_into_xml(self, since):
        count, cpvs = self.get_wellknown_cpvs(since)

        # nothing to do
        if count == 0:
            return None, self._out

        # first is the temp dir the files are stored in
        categoryFiles = []
        categoryFiles.append(mkdtemp(None, 'pfl-'))

        workingon = 0
        for r in cpvs: # repository
            for c in cpvs[r]: # category
                # https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp
                categoryFile = mkstemp('.xml', 'pfl-', categoryFiles[0])
                self._write2file(categoryFile, '<?xml version="1.0" encoding="UTF-8"?>', '')
                self._write2file(categoryFile, '<pfl xmlns="http://www.portagefilelist.de/xsd/collect">', '\n')
                self._write2file(categoryFile, '<category name="%s">' % c, '\n\t')
                categoryFiles.append(categoryFile[1])

                for p in cpvs[r][c]: # packages
                    for v in cpvs[r][c][p]: # versions
                        workingon = workingon + 1
                        self.log('working on (%d of %d) %s/%s-%s::%s' % (workingon, count, c, p, v, r))

                        contents = self.get_contents(c, p, v)

                        # no files -> this package does not matter
                        if len(contents) == 0:
                            continue

                        mergedstamp = self._vardbapi.aux_get('%s/%s-%s' % (c, p, v), ['_mtime_'])[0]
                        # Some installs started to create float timestamp
                        # I don't know why it was not float already all the time
                        # Something in the creation of _mtime_ must have been changed
                        mergedstamp = int(round(mergedstamp))

                        use = self._vardbapi.aux_get('%s/%s-%s' % (c, p, v), ['USE'])[0].split()
                        iuse = self._vardbapi.aux_get('%s/%s-%s' % (c, p, v), ['IUSE'])[0].split()
                        keywords = self._vardbapi.aux_get('%s/%s-%s' % (c, p, v), ['KEYWORDS'])[0].split()

                        us = []
                        arch = 'none'

                        for u in use:
                            if u in iuse:
                                us.append(u)
                            if u in keywords or '~' + u in keywords:
                                arch = u
                        self._write2file(categoryFile, '<package arch="%s" name="%s" timestamp="%s" version="%s" repo="%s">' % (arch, p, str(mergedstamp), v, r), '\n\t\t')

                        self._write2file(categoryFile, '<files>', '\n\t\t\t')
                        for f in contents:
                            self._write2file(categoryFile, '<file type="%s">%s</file>' % (contents[f][0], escape(f)), '\n\t\t\t\t')
                        self._write2file(categoryFile, '</files>', '\n\t\t\t')

                        if len(us) > 0:
                            self._write2file(categoryFile, '<uses>', '\n\t\t\t')
                            for u in us:
                                self._write2file(categoryFile, '<use>%s</use>' %u, '\n\t\t\t\t')
                            self._write2file(categoryFile, '</uses>', '\n\t\t\t')

                        self._write2file(categoryFile, '</package>', '\n\t\t')
                self._write2file(categoryFile, '</category>', '\n\t')
                self._write2file(categoryFile, '</pfl>', '\n')
                os.close(categoryFile[0])

        return categoryFiles, self._out


class PFL(object):
    _lastrun = 0
    _config = None
    _options = None
    _out = ''

    def __init__(self, options):
        self._options = options
        self._read_config()

    def _finish(self, xmlfiles, success = True):
        if success and not self._options['pretend']:
            if not self._config.has_section('PFL'):
                self._config.add_section('PFL')

            self._config.set('PFL', 'lastrun', str(int(time())))
            self._config.set('PFL', 'version', VERSION)

            hconfig = open(INFOFILE, 'w')
            self._config.write(hconfig)
            hconfig.close()

        if xmlfiles and os.path.isdir(xmlfiles[0]):
            if self._options['pretend']:
                self.log('Pretend mode. Keeping:\n'+'\n'.join(xmlfiles))
                self.log('The files need to be removed manually!')
            else:
                self.log('Cleanup ...')
                # the folder is the first element
                tmpDir = xmlfiles.pop(0)
                self.log(tmpDir + '*')
                for pathToBeRemoved in xmlfiles:
                    os.unlink(pathToBeRemoved)
                os.rmdir(tmpDir)
                self.log('Done.')

    def _read_config(self):
        self._config = configparser.ConfigParser()
        if os.path.isfile(INFOFILE):
            self._config.read(INFOFILE)

    def _last_run(self):
        if self._config.get('PFL', 'version', fallback='noversion') == 'noversion':
            return 0
        else:
            return int(self._config.get('PFL', 'lastrun', fallback=0))

    def log(self, output):
        if self._options['stdout']:
            print(output)
        else:
            self._out += output + '\n'

    def run(self):
        pm = PortageMangle(self._options)

        xmlfiles, msg = pm.collect_into_xml(self._last_run())
        self.log(msg)

        if xmlfiles == None:
            self.log('Nothing to collect. If this is wrong, set PFL/lastrun in {} to 0'.format(INFOFILE))
            toClean = xmlfiles
        elif self._options['pretend']:
            self.log('Pretend mode. Nothing to upload.')
            toClean = xmlfiles
        else:
            toClean = []
            tmpDir = xmlfiles.pop(0)
            toClean.append(tmpDir)

            self.log('Creating file to upload ...')
            fileToUpload = tmpDir + '.tar'
            tarFile = tarfile.open(fileToUpload, 'w')
            for toBeCompressed in xmlfiles:
                os.system('bzip2 %s' % toBeCompressed)
                fPath = toBeCompressed + '.bz2'
                tarFile.add(fPath, arcname=os.path.basename(fPath))
                toClean.append(fPath)
            tarFile.close()
            self.log('Done: {}'.format(fileToUpload))
            toClean.append(fileToUpload)

            try:
                self.log('Uploading file {} to {} ...'.format(fileToUpload, UPLOADURL))
                files = {'foo': open(fileToUpload, 'rb')}
                r = requests.post(UPLOADURL, files=files)
                if r.status_code == 200:
                    self.log('Upload successful')
                    self.log('{}'.format(r.text))
                else:
                    self.log('Error: Upload was not successful')
                    self.log('HTTP Response Code: {}'.format(r.status_code))
                    self.log('HTTP Response Body: {}'.format(r.text))
                    self._finish(toClean, False)
                    return 1, self._out
            except Exception as e:
                sys.stderr.write("%s\n" % e)
                self._finish(toClean, False)
                return 1, self._out

        self._finish(toClean, True)
        return 0, self._out
